inc/fftw.hpp

```diff
@@ -0,0 +1,629 @@
+
+      void create() {
+	//TODO: set number of threads
+      
+      }
+
+      void destroy() {
+      
+	//not needed
+      }
+    };
+
+
+    /**
+     * Fftw plan and execution class.
+     *
+     * This class handles:
+     * - {1D, 2D, 3D} x {R2C, C2R, C2C} x {inplace, outplace} x {float, double}.
+     */
+    template<typename TFFT, // see fft_abstract.hpp (FFT_Inplace_Real, ...)
+	     typename TPrecision, // double, float
+	     size_t   NDim // 1..3
+	     >
+    struct FftwImpl
+    {
+
+      //////////////////////////////////////////////////////////////////////////////////////
+      // COMPILE TIME FIELDS
+    
+      using Extent = std::array<std::size_t,(int)NDim>;
+      using Api  = typename traits::plan<TPrecision>;
+      using ComplexType = typename traits::plan<TPrecision>::ComplexType;
+      using RealType = typename traits::plan<TPrecision>::RealType;
+      using PlanType = typename traits::plan<TPrecision>::PlanType;
+    
+      // using FFTExecuteForward  = typename traits::plan::FFTExecuteForward;
+      // using FFTExecuteBackward = typename traits::plan::FFTExecuteBackward;
+
+      static constexpr
+      bool IsInplace = TFFT::IsInplace;
+      static constexpr
+      bool IsComplex = TFFT::IsComplex;
+      static constexpr
+      bool Padding = IsInplace && IsComplex==false && NDim>1;
+
+      using value_type  = typename std::conditional<IsComplex,ComplexType,RealType>::type;
+    
+      // static constexpr
+      //  fftwType FFTForward  = IsComplex ? traits::plan::FFTComplex : traits::plan::FFTForward;
+      // static constexpr
+      //  fftwType FFTBackward = IsComplex ? traits::plan::FFTComplex : traits::plan::FFTBackward;
+
+      //////////////////////////////////////////////////////////////////////////////////////
+      // RUNTIME TIME FIELDS
+    
+      size_t n_;        // =[1]*..*[dim]
+      size_t n_padded_; // =[1]*..*[dim-1]*([dim]/2+1)
+      Extent extents_;
+
+      PlanType			plan_           ;
+      value_type*		data_           = nullptr;
+      ComplexType*       	data_transform_ = nullptr; // intermediate buffer
+      size_t             	data_size_;
+      size_t             	data_transform_size_;
+
+      FftwImpl(const Extent& cextents)
+	: n_(0),
+	  n_padded_(0),
+	  extents_(),
+	  plan_(),
+	  data_(nullptr),
+	  data_transform_(nullptr),
+	  data_size_(0),
+	  data_transform_size_(0)
+      {
+	extents_ = interpret_as::column_major(cextents);
+	
+	n_ = std::accumulate(extents_.begin(),
+			     extents_.end(),
+			     1,
+			     std::multiplies<unsigned>());
+	if(Padding)
+	  n_padded_ = n_ / extents_[NDim-1] * (extents_[NDim-1]/2 + 1);
+
+	data_size_ = ( Padding ? 2*n_padded_ : n_ ) * sizeof(value_type);
+	data_transform_size_ = IsInplace ? 0 : n_ * sizeof(ComplexType);
+
+	traits::thread_api<TPrecision>::init_threads();
+	traits::thread_api<TPrecision>::plan_with_threads(-1);
+      }
+
+      ~FftwImpl(){
+	traits::thread_api<TPrecision>::cleanup_threads();
+      }
```

Comment:
looking at it in more detail, I have to say that I can't move the thread init calls to the context as it requires the precision to be known. fftw is so "well" written that initializing threads depends on the type to be used!! aka there is fftwf_init_threads for float32_t and there is fftw_init_threads for float64_t data! we can also make the context be TPrecision aware but that may result in more refactoring in the other FFT benchmarks. what do you think?